File: docs/trackhub-schema/validate.pl
#!/usr/bin/env perl

use strict;
use warnings;

use JSON;
use JSON::Schema;
use Data::Dumper;

my ($schema_file, $json_file) = @ARGV;

my ($schema_string, $json_string) = 
  (&slurp_file($schema_file), &slurp_file($json_file));

my $validator = JSON::Schema->new($schema_string); #, %options);
my $json      = from_json($json_string);
my $result    = $validator->validate($json);
 
if ($result) {
  print "Valid!\n";
} else {
  print "Errors\n";
  print " - $_\n" foreach $result->errors;
}

sub slurp_file {
  my $file = shift;

  my $string;
  {
    local $/=undef;
    open FILE, "<$file" or die "Couldn't open file: $!";
    $string = <FILE>;
    close FILE;
  }
  
  return $string;
}

File: docs/trackhub-schema/validate.py
#!/usr/bin/env python

import sys
import getopt
import json
from jsonschema import Draft4Validator
# from pprint import pprint

def main(argv):
    (schema, instance) = ('','')
    try:                                
        opts, args = getopt.getopt(argv, "hs:f:")
    except getopt.GetoptError:          
        usage()                         
        sys.exit(2)                     
    for opt, arg in opts:
        if opt == '-h':
            usage()     
            sys.exit()
        elif opt == '-s':
            with open(arg) as json_file:
                schema = json.load(json_file)
        elif opt == '-f':
            with open(arg) as json_file:
                instance = json.load(json_file)

    if not schema or not instance:
        usage()
        sys.exit(2)

    v = Draft4Validator(schema)
    errors = sorted(v.iter_errors(instance), key=lambda e: e.path)
    for error in errors:
        print(error)
        # print(error.message)
        # print(list(error.path))
        for suberror in sorted(error.context, key=lambda e: e.schema_path):
            print(list(suberror.schema_path), suberror.message)    


def usage():
    print "validate.py\n[options]\n\t-s <schema>\n\t-f <file>\n"

if __name__ == "__main__":
    main(sys.argv[1:])
